feat: course dates

mumer92 opened the pull request:
This PR adds support for Course Dates. https://github.com/openedx/openedx-app-ios/issues/85

This is preliminary PR with the Course Dates feature. Designs are TBD.

Screenshots:
<img width="180" alt="Screenshot 2023-10-31 at 11 56 18 PM" src="https://github.com/openedx/openedx-app-ios/assets/960241/6df4b03a-1925-49ad-997e-d11012830fab">
<img width="175" alt="Screenshot 2023-10-31 at 11 56 35 PM" src="https://github.com/openedx/openedx-app-ios/assets/960241/6cefc91a-50a9-40f2-bf64-8b898597fa51">



_Note_
Not included in this ticket:
- Calendar integration
- PLS (shift due dates)
- Linking of items on the dates page to specific assessments within the app

volodymyr-chekyrta commented on Course/Course/Domain/Model/CourseDates.swift:
```diff
@@ -0,0 +1,210 @@
+                description: "",
+                learnerHasAccess: true,
+                link: "", linkText: nil,
+                title: "Today",
```
It would be great to move this to string resources for translations.

volodymyr-chekyrta commented on Course/Course/Domain/Model/CourseDates.swift:
```diff
@@ -0,0 +1,210 @@
+        }
+    }
+    
+    var title: String {
+        switch self {
+        case .completed:
+            return "Completed"
+        case .pastDue:
+            return "Past due"
+        case .dueNext:
+            return "Due next"
+        case .unreleased:
+            return "Unreleased"
+        case .verifiedOnly:
+            return "Verified Only"
```
Same to these.
It would be great to move them to string resources for translations.

volodymyr-chekyrta commented on Core/Core.xcodeproj.xcworkspace/xcuserdata/m.umer.xcuserdatad/UserInterfaceState.xcuserstate:
Seems like .gitignore issue, xcuserdata/m.umer.xcuserdatad folder should be ignored

volodymyr-chekyrta commented on Course/Course/Presentation/Dates/CourseDatesView.swift:
```diff
@@ -0,0 +1,279 @@
+    func getStatusDetails(for blockStatus: BlockStatus) -> (String, Color, Color) {
+        switch blockStatus {
+        case .verifiedOnly:
+            return ("Verified Only", Color.white, Color.black.opacity(0.5))
+        case .pastDue:
+            return ("Past Due", Color.black, Color.gray.opacity(0.4))
+        case .dueNext:
+            return ("Due Next", Color.white, Color.black.opacity(0.5))
+        case .unreleased:
+            return ("Unreleased", Color.white.opacity(0), Color.white.opacity(0))
```
Let's also move these strings to the Localizable.strings for translations

volodymyr-chekyrta commented on Course/Course/Presentation/Dates/CourseDatesViewModel.swift:
```diff
@@ -0,0 +1,72 @@
+        self.connectivity = connectivity
+    }
+        
+    public var sortedDates: [Date] {
+        courseDates?.sortedDateToCourseDateBlockDict.keys.sorted() ?? []
+    }
+    
+    public func blocks(for date: Date) -> [CourseDateBlock] {
+        courseDates?.sortedDateToCourseDateBlockDict[date] ?? []
+    }
```
Unnecessary public level modifier if we use it only in the Course module

volodymyr-chekyrta commented on Core/Core.xcodeproj/project.xcworkspace/xcuserdata/m.umer.xcuserdatad/UserInterfaceState.xcuserstate:
Seems like .gitignore issue, xcuserdata/m.umer.xcuserdatad folder should be ignored

volodymyr-chekyrta commented on Course/Course.xcodeproj/project.xcworkspace/xcuserdata/m.umer.xcuserdatad/UserInterfaceState.xcuserstate:
Seems like .gitignore issue, xcuserdata/m.umer.xcuserdatad folder should be ignored

volodymyr-chekyrta commented on Discovery/Discovery.xcodeproj/project.xcworkspace/xcuserdata/m.umer.xcuserdatad/UserInterfaceState.xcuserstate:
Seems like .gitignore issue, xcuserdata/m.umer.xcuserdatad folder should be ignored

mumer92 commented:
@volodymyr-chekyrta i have updated the PR, added translations and added few files to .gitignore to ignore the files that were previously checked into the repo.

volodymyr-chekyrta commented:
@mumer92 Thank you! 🙌

Would you mind to check the `CourseDateBlock` file? 
I noticed that `var blockTitle: String` appears to still be hardcoded. 

Could you also check those files? I'm uncertain, but it seems to me that they may still be present in the project:
`
Core/Core.xcodeproj.xcworkspace/xcuserdata/m.umer.xcuserdatad
Course/Course.xcodeproj/project.xcworkspace/xcuserdata/m.umer.xcuserdatad
Discovery/Discovery.xcodeproj/project.xcworkspace/xcuserdata/m.umer.xcuserdatad
`

eyatsenkoperpetio commented:
Hi, small problem @mumer92 
<img width="408" alt="Screenshot 2023-11-03 at 15 48 59" src="https://github.com/openedx/openedx-app-ios/assets/76485998/a4a9cb49-4c41-4ad5-ae6a-68ceff4c47ba">

mumer92 commented:
@eyatsenkoperpetio i have created https://github.com/openedx/openedx-app-ios/pull/149

touchapp commented:
@marcotuts can we have Sam take a look at this? I think it needs some design love.
